build-tools/packages/build-tools/src/common/biome2Config.ts

```diff
@@ -0,0 +1,523 @@
+ * but the different type signatures (BiomeConfigRaw vs Biome2ConfigRaw) and opaque return types
+ * provide compile-time guarantees that prevent mixing 1.x and 2.x configs accidentally.
+ *
+ * @param configPaths - Array of absolute paths to config files, ordered from base to most specific.
+ *                      Each config's values will override those from earlier configs.
+ * @returns The merged configuration.
+ */
+async function mergeMultipleBiome2Configs(
+	configPaths: string[],
+): Promise<Biome2ConfigResolved> {
+	const allConfigs = await Promise.all(
+		configPaths.map((pathToConfig) => loadRawBiome2Config(pathToConfig)),
+	);
+
+	const mergedConfig = merge.withOptions(
+		{
+			// Biome does not merge arrays
+			mergeArrays: false,
+		},
+		...allConfigs,
+	);
+
+	return mergedConfig as Biome2ConfigResolved;
+}
+
+type Biome2ConfigSection = "formatter" | "linter";
+
+/**
+ * Given a Biome 2.x config object, returns the ordered patterns from 'includes' across the 'files'
+ * and the specified section ('formatter' or 'linter') in the config.
+ *
+ * Patterns are returned in declaration order, which is essential for correct re-inclusion behavior.
+ * In Biome 2.x, patterns like `["**", "!test/**", "test/special/**"]` are processed in order:
+ * 1. Include all files (**)
+ * 2. Exclude test/** (!test/**)
+ * 3. Re-include test/special/** (test/special/**)
+ *
+ * See: {@link https://biomejs.dev/reference/configuration/#filesinclude}
+ *
+ * @param config - A resolved/merged Biome 2.x configuration.
+ * @param section - The config section to extract patterns from ('formatter' or 'linter').
+ * @returns An array of patterns in declaration order, with negation prefixes preserved.
+ */
+export function getOrderedPatternsFromBiome2Config(
+	config: Biome2ConfigResolved,
+	section: Biome2ConfigSection,
+): string[] {
+	const patterns: string[] = [];
+
+	// Add patterns from files.includes first
+	if (config.files?.includes) {
+		patterns.push(...config.files.includes);
+	}
+
+	// Add section-specific patterns
+	const sectionConfig = config[section];
+	if (sectionConfig?.includes) {
+		patterns.push(...sectionConfig.includes);
+	}
+
+	return patterns;
+}
+
+/**
+ * Return an array of absolute paths to files that Biome 2.x would format under the provided path. Note that .gitignored
+ * paths are always excluded, regardless of the "vcs" setting in the Biome configuration.
+ *
+ * @param directoryOrConfigFile - A path to a directory or a Biome config file. If a directory is provided, then the
+ * closest Biome configuration will be loaded and used. If a path to a file is provided, it is assumed to be a Biome
+ * config file and will be loaded as such. The directory containing the config file will be used as the working
+ * directory when applying the Biome include/ignore settings.
+ * @param gitRepo - A GitRepo instance that is used to enumerate files.
+ */
+export async function getBiome2FormattedFilesFromDirectory(
```

Comment:
What is the difference between this function and the getBiome2FormattedFiles function? Why are both needed?